File: e2e/elixir/test/robots_test.exs
# E2e tests for category: robots
defmodule E2e.RobotsTest do
  use ExUnit.Case, async: true

  describe "robots_allow_all" do
    test "Permissive robots.txt allows all paths" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.robots.is_allowed) == true
    end
  end

  describe "robots_allow_override" do
    test "Allow directive overrides Disallow for specific paths" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.robots.is_allowed) == true
    end
  end

  describe "robots_comments_handling" do
    test "Correctly parses robots.txt with inline and line comments" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.robots.is_allowed) == true
    end
  end

  describe "robots_crawl_delay" do
    test "Respects crawl-delay directive from robots.txt" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.robots.crawl_delay) == 2
    end
  end

  describe "robots_disallow_path" do
    test "Robots.txt disallows specific paths" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.robots.is_allowed) == false
    end
  end

  describe "robots_meta_nofollow" do
    test "Detects nofollow meta robots tag and skips link extraction" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.robots.nofollow_detected) == true
    end
  end

  describe "robots_meta_noindex" do
    test "Detects noindex meta robots tag in HTML page" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.robots.noindex_detected) == true
    end
  end

  describe "robots_missing_404" do
    test "Missing robots.txt (404) allows all crawling" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.robots.is_allowed) == true
    end
  end

  describe "robots_multiple_user_agents" do
    test "Picks the most specific user-agent block from robots.txt" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.robots.is_allowed) == true
    end
  end

  describe "robots_request_rate" do
    test "Parses request-rate directive from robots.txt" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.robots.crawl_delay) == 5
      assert String.trim(result.robots.is_allowed) == true
    end
  end

  describe "robots_sitemap_directive" do
    test "Discovers sitemap URL from Sitemap directive in robots.txt" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.robots.is_allowed) == true
    end
  end

  describe "robots_user_agent_specific" do
    test "Matches user-agent specific rules in robots.txt" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.robots.is_allowed) == false
    end
  end

  describe "robots_wildcard_paths" do
    test "Handles wildcard Disallow patterns in robots.txt" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.robots.is_allowed) == false
    end
  end

  describe "robots_x_robots_tag" do
    test "Respects X-Robots-Tag HTTP header directives" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.robots.x_robots_tag) == "noindex, nofollow"
      assert String.trim(result.robots.noindex_detected) == true
      assert String.trim(result.robots.nofollow_detected) == true
    end
  end
end

File: e2e/rust/tests/robots_test.rs
//! E2e tests for category: robots

use kreuzcrawl::scrape;
use kreuzcrawl::create_engine;

#[tokio::test]
async fn test_robots_allow_all() {
    // Permissive robots.txt allows all paths
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert_eq!(result.is_allowed, true, "equals assertion failed");
}

#[tokio::test]
async fn test_robots_allow_override() {
    // Allow directive overrides Disallow for specific paths
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert_eq!(result.is_allowed, true, "equals assertion failed");
}

#[tokio::test]
async fn test_robots_comments_handling() {
    // Correctly parses robots.txt with inline and line comments
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert_eq!(result.is_allowed, true, "equals assertion failed");
}

#[tokio::test]
async fn test_robots_crawl_delay() {
    // Respects crawl-delay directive from robots.txt
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    let crawl_delay = result.crawl_delay.as_deref().unwrap_or("");
    assert_eq!(crawl_delay, 2, "equals assertion failed");
}

#[tokio::test]
async fn test_robots_disallow_path() {
    // Robots.txt disallows specific paths
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert_eq!(result.is_allowed, false, "equals assertion failed");
}

#[tokio::test]
async fn test_robots_meta_nofollow() {
    // Detects nofollow meta robots tag and skips link extraction
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert_eq!(result.nofollow_detected, true, "equals assertion failed");
}

#[tokio::test]
async fn test_robots_meta_noindex() {
    // Detects noindex meta robots tag in HTML page
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert_eq!(result.noindex_detected, true, "equals assertion failed");
}

#[tokio::test]
async fn test_robots_missing_404() {
    // Missing robots.txt (404) allows all crawling
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert_eq!(result.is_allowed, true, "equals assertion failed");
}

#[tokio::test]
async fn test_robots_multiple_user_agents() {
    // Picks the most specific user-agent block from robots.txt
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert_eq!(result.is_allowed, true, "equals assertion failed");
}

#[tokio::test]
async fn test_robots_request_rate() {
    // Parses request-rate directive from robots.txt
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    let crawl_delay = result.crawl_delay.as_deref().unwrap_or("");
    assert_eq!(crawl_delay, 5, "equals assertion failed");
    assert_eq!(result.is_allowed, true, "equals assertion failed");
}

#[tokio::test]
async fn test_robots_sitemap_directive() {
    // Discovers sitemap URL from Sitemap directive in robots.txt
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert_eq!(result.is_allowed, true, "equals assertion failed");
}

#[tokio::test]
async fn test_robots_user_agent_specific() {
    // Matches user-agent specific rules in robots.txt
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert_eq!(result.is_allowed, false, "equals assertion failed");
}

#[tokio::test]
async fn test_robots_wildcard_paths() {
    // Handles wildcard Disallow patterns in robots.txt
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert_eq!(result.is_allowed, false, "equals assertion failed");
}

#[tokio::test]
async fn test_robots_x_robots_tag() {
    // Respects X-Robots-Tag HTTP header directives
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    let x_robots_tag = result.x_robots_tag.as_deref().unwrap_or("");
    assert_eq!(x_robots_tag.trim(), r#"noindex, nofollow"#, "equals assertion failed");
    assert_eq!(result.noindex_detected, true, "equals assertion failed");
    assert_eq!(result.nofollow_detected, true, "equals assertion failed");
}
